packages/features/src/web-connector/routes/web-connector.tsx

```diff
@@ -8,42 +9,151 @@ import yaml from "yaml"
   return xss(yamlPayload)
 }
 
+const sendSandboxMessage = (payload: Json) => {
+  const sandbox = document.querySelector(
+    "#o1sandbox",
+  ) as HTMLIFrameElement | null
+  if (!sandbox) return
+  return sandbox.contentWindow?.postMessage(payload, "*")
+}
+
 type ActionRequest = {
   title: string
   submitButtonLabel?: string
   rejectButtonLabel?: string
   payload: string
-  inputType: "text" | "password" | "confirmation"
-  loading: boolean
+  inputType: "text" | "password" | "confirmation" | "selection"
+  contract: string | undefined
   emitConnected: boolean
 }
 
 export const WebConnectorRoute = () => {
+  const [loading, setLoading] = useState(true)
+  const [loadingMessage, setLoadingMessage] = useState<string>()
   const [request, setRequest] = useState<ActionRequest>({
     title: "",
     payload: "",
     inputType: "confirmation",
-    loading: true,
+    contract: undefined,
     emitConnected: false,
   })
-  const onSubmit: SubmitHandler<UserInputForm> = async ({ userInput }) => {
+  const startSubmitting: SubmitHandler<UserInputForm> = async ({
+    userInput,
+  }) => {
+    if (!request.contract) return onSubmit({ userInput })
+    setLoading(true)
+    try {
+      // For presentation contract, set up message handler for signing request
+      if (request.contract === "presentation") {
+        setLoadingMessage("Creating presentation...")
+        const messageHandler = async (event: MessageEvent) => {
+          if (event.data.type === "presentation-signing-request") {
+            try {
+              // Use the passphrase we already have to sign fields
+              const response = (await sendMessage(
+                "mina_signFieldsWithPassphrase",
+                {
+                  params: [
+                    {
+                      fields: event.data.fields as string[],
+                      passphrase: userInput,
+                    },
+                  ],
+                  context: {
+                    origin: `chrome-extension://${chrome.runtime.id}`,
+                  },
+                },
+                "background",
+              )) as SignedFieldsResponse
+
+              if (!response || !response.signature) {
+                throw new Error(
+                  `Failed to sign fields; response: ${JSON.stringify(response)}`,
+                )
+              }
+
+              // Send signature back to sandbox
+              sendSandboxMessage({
+                type: "presentation-signature",
+                signature: response.signature,
+              })
+            } catch (error) {
+              console.error("Error signing fields:", error)
+              sendSandboxMessage({
+                type: "presentation-signature-error",
+                error: error instanceof Error ? error.message : "Unknown error",
+              })
+            }
+          }
+        }
+
+        window.addEventListener("message", messageHandler)
```

Comment:
who can send messages to this listener? is this isolated from the web page? I'm just worried the web page could trick the wallet into signing something

maybe that's a question for @mrcnk

Reply:
Yeah, especially that we have `signFieldsWithPassphrase`, I would check if we can change `window` message handler. If it's to communicate with iframe, we can check `iframe.contentWindow.body.addEventListener`. If we need to rely on window, we should check if there are suitable methods in `webext-bridge` that we use for messaging in Pallad.